jetty-core/jetty-start/src/main/java/org/eclipse/jetty/start/CommandLineBuilder.java

```diff
@@ -38,31 +38,82 @@ public static String findJavaBin()
         return "java";
     }
 
-    /**
-     * Perform an optional quoting of the argument, being intelligent with spaces and quotes as needed. If a subString is set in quotes it won't the subString
-     * won't be escaped.
-     *
-     * @param arg the argument to quote
-     * @return the quoted and escaped argument
-     * @deprecated no replacement, quoting is done by {@link #toQuotedString()} now.
-     */
-    @Deprecated
-    public static String quote(String arg)
+    private final StringBuilder commandLine = new StringBuilder();
+    private final List<String> args = new ArrayList<>();
+    private final String separator;
+
+    public CommandLineBuilder()
     {
-        return "'" + arg + "'";
+        this(false);
     }
 
-    private List<String> args;
-
-    public CommandLineBuilder()
+    public CommandLineBuilder(boolean multiline)
     {
-        args = new ArrayList<String>();
+        separator = multiline ? (" \\" + System.lineSeparator() + "  ") : " ";
     }
 
-    public CommandLineBuilder(String bin)
+    /**
+     * Quote a string suitable for use with a command line shell using double quotes.
+     * <p>This method applies doubles quoting as described for the unix {@code sh} commands:
+     * Enclosing characters within double quotes preserves the literal meaning of all characters except
+     * dollarsign ($), backquote (`), and backslash (\).
+     * The backslash inside double quotes is historically weird, and serves
+     * to quote only the following characters: {@code $ ` " \ newline}.
+     * Otherwise it remains literal.
+     * </p><p>Additionally, a string is deemed to need quoting if
+     * it contains a single quote or a {@code bash} meta character: A character that,
+     * when unquoted, separates words. One of the following: {@code |  & ; ( ) < > space tab newline}
+     * </p>
+     *
+     * @param input The string to quote if needed
+     * @return The quoted string or the original string if quotes are not necessary
+     */
+    public static String shellQuoteIfNeeded(String input)
```

Comment:
The list of characters for bash (haven't looked at other shells yet) that have special meaning ...

Character | Meaning
----- | -----
`~` | home directory
<code>&#x60;</code> | Single quote for command substitution
`#` | comment
`$` | variable
`&` | background
`*` | string wildcard
`(` | start substitution
`)` | end substitution
`\` | quote next char
<code>&#x7C;</code> | pipe to other command
`[` | start character-set wildcard
`]` | end character-set wildcard
`{` | start command block
`}` | end command block
`;` | shell command separator
`'` | strong quote
`"` | weak quote
`<` | input redirection
`>` | output redirection
`/` | pathname separator
`?` | single character wildcard
`!` | pipeline logical not operator

Each of those should trigger quoting and in some cases escaping

Reply:
@joakime I think we don't need every special character, we just need metacharacters that are used to split a command line and characters involved with quoting itself..... 
Or do we need to also protect wild cards? Hmmm probably yes....

Reply:
I wonder if it would be simpler to have a set of characters that we don't need to quote for: a-z A-Z 0-9 _-/

Reply:
Here's a longer breakdown of special characters for the shells: csh, sh, bash, ksh, zsh



Character(s) | Shells | Meaning
----- | ----- | -----
`~` | sh, csh, bash, ksh, zsh | home directory
<code>&#x60;</code> | bash |  Single quote for command substitution
`#` | csh, sh, bash | comment
`#` | ksh, bash | variable replacement
`^` | sh _only_ | pipe character
`^` | csh, bash | previous command
`$` | csh, sh, bash, ksh, zsh | variable
`&` | csh, sh, bash, ksh, zsh | background job. or combine stdin and stdout
`*` | csh, sh, bash | string wildcard (match zero or more characters)
`(` `)` | bash, ksh | command substitution
`\` | csh, sh, bash, ksh, zsh | quote next char
<code>&#x7C;</code> | csh, sh, bash, ksh | pipe to other command
<code>&#x7C;</code> | zsh | job control
`[` `]` | csh, sh, bash, ksh | character-set wildcard (match range of characters)
`{` `}` | csh, bash | in-line expansion
`{` `}` | csh, bash, ksh, zsh | command block
`:` | sh | evaluate args. or separate values in paths.
`:` | csh | variable modifier
`;` | csh, sh, bash, ksh, zsh | shell command separator
`;` | sh | end of _case_ element
`'` | bash | strong quote
`"` | csh, sh, bash | weak quote
`<` | csh, sh, bash, ksh, zsh | input redirection
`>` | csh, sh, bash, ksh, zsh | output redirection
`/` | csh, sh, bash, ksh, zsh | pathname separator
`?` | csh, sh, bash | single character wildcard
`?` | sh, bash | optional variable
`!` | csh, bash | pipeline logical not operator
`!` | ksh, zsh | exit status

Reply:
> I wonder if it would be simpler to have a set of characters that we don't need to quote for: a-z A-Z 0-9 _-/

I wouldn't include `-` (dash) or `/` (slash) in that list.

The `-` (dash) is used as stdin or stdout reference in many shells.
And the `/` (slash) is used as a hard path separator in some shells (forces the shell to intepret the text around the `/` in path terms, applying path behaviors to that string), and a soft separator in others (if present, the matching logic applies to the strings adjacent too)

Reply:
For Windows, the rules are a bit different ...

You can only use the `\` back-slash to escape non-control characters when executing in CMD (such as a space character).

In Powershell, the escaping is done with `^` character.

We will likely need separate quoting and escaping methods for Unix/Linux/OSX and Windows. (Maybe even a special method for those users on Powershell)

Reply:
I think this can only really be output for unix shells, and then only the common ones.   It is a tool for human debugging and for doing start scripts.   
If we keep the output constant, then others can transform to any special usages

Reply:
I've updated the `usage.txt` file to make it clear that the output is only suitable for the unix `sh` command.  I do not think we can come up with an output format that will work on every shell on every platform.  I also think that any attempt to cover multiple platforms is just going to result in hard to handle variable output.  Best to just give fixed output and allows any users on other platforms either to use a sh port or do a transformation.